kie-soup-maven-utils/kie-soup-maven-integration/src/main/java/org/appformer/maven/integration/InJarArtifactResolver.java

```diff
@@ -0,0 +1,169 @@
+import java.util.Optional;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+import org.apache.maven.project.MavenProject;
+import org.appformer.maven.integration.embedder.EmbeddedPomParser;
+import org.appformer.maven.integration.embedder.MavenProjectLoader;
+import org.appformer.maven.support.AFReleaseId;
+import org.appformer.maven.support.AFReleaseIdImpl;
+import org.appformer.maven.support.DependencyFilter;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.artifact.DefaultArtifact;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+public class InJarArtifactResolver extends ArtifactResolver {
+
+    private static final Logger log = LoggerFactory.getLogger(InJarArtifactResolver.class);
+
+    private ClassLoader classLoader;
+    private List<URL> jarRepository;
+    private PomParser pomParser;
+
+    InJarArtifactResolver(ClassLoader classLoader, AFReleaseId releaseId) {
+        this.classLoader = classLoader;
+        this.jarRepository = new ArrayList<>();
+        init(releaseId);
+    }
+
+    public boolean isLoaded() {
+        return pomParser != null;
+    }
+    // initialize in jar repository
+    private void init(AFReleaseId releaseId) {
+        jarRepository = buildResources();
+        pomParser = buildPomParser(releaseId);
+    }
+
+    private List<URL> buildResources() {
+        URL resourceURL = this.classLoader.getResource("");
+        if (resourceURL == null) {
+            return emptyList();
+        }
+        List<URL> resources = new ArrayList<>();
+        try (InputStream is = resourceURL.openStream();
+                ZipInputStream stream = new ZipInputStream(is);) {
+
+            ZipEntry entry = null;
+            while ((entry = stream.getNextEntry()) != null) {
+                if (isInJarFolder(entry.getName())) {
+                    resources.add(classLoader.getResource(entry.getName()));
+                }
+            }
+            if (!resources.isEmpty()) {
+                log.info("Found in jar repository {}", jarRepository);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return resources;
+    }
+
+    private PomParser buildPomParser(AFReleaseId releaseId) {
+        try {
+            List<URL> url = jarRepository.stream().filter(e -> e.getFile().endsWith(toFile(releaseId))).collect(toList());
+            if (url.isEmpty()) {
+                return null;
+            }
+            URL pomFile = classLoader.getResource(url.get(0) + "!/META-INF/maven/" + releaseId.getGroupId() + "/" + releaseId.getArtifactId() + "/pom.xml");
+            InputStream pomStream = pomFile.openStream();
+            MavenProject mavenProject = MavenProjectLoader.parseMavenPom(pomStream);
+            PomParser artifactPomParser = new EmbeddedPomParser(mavenProject);
+            pomStream.close();
```

Comment:
Shouldn't it be in a finally clause? `parseMavenPom` method may throw a _RuntimeException_, and therefore stream wouldn't be closed.

Reply:
moved to a try-witth-resources block